backend/src/rate_limiter/provider.rs

```diff
@@ -0,0 +1,164 @@
+use async_trait::async_trait;
+use sqlx::PgPool;
+
+use super::{Allowance, RLKey, StorageProvider};
+
+pub(crate) struct RLPostgresStorage {
+    pool: PgPool,
+}
+
+impl RLPostgresStorage {
+    pub(crate) fn new(pool: PgPool) -> Self {
+        RLPostgresStorage { pool }
+    }
+}
+
+#[async_trait]
+impl StorageProvider for RLPostgresStorage {
+    async fn get(&self, key: &RLKey) -> Option<Allowance> {
+        Allowance::load(&self.pool, key).await.ok().flatten()
+    }
+
+    async fn set(&self, key: RLKey, allowance: Allowance) {
+        if let Err(e) = allowance.store(&self.pool, &key).await {
```

Comment:
Minor: s/e/error, then the structured logging will have a better field description.